wardrobe/tradable_kitties/src/lib.rs

```diff
@@ -0,0 +1,446 @@
+                    let price = match td_input_kitty.price {
+                        None => return Err(TradeableKittyError::KittyNotForSale),
+                        Some(p) => p,
+                    };
+                    total_price_of_kitty = total_price_of_kitty
+                        .checked_add(price)
+                        .ok_or(TradeableKittyError::MoneyError(MoneyError::ValueOverflow))?;
+                }
+                Some(_) => {
+                    // More than 1 tradable kitty are sent for trading.
+                    return Err(TradeableKittyError::CannotBuyMoreThanOneKittyAtTime);
+                }
+            };
+        } else {
+            return Err(TradeableKittyError::BadlyTyped);
+        }
+    }
+
+    // Ensuring we found kitty in the input
+    ensure!(input_ktty_to_be_traded != None, {
+        TradeableKittyError::InputMissingError
+    });
+
+    // Seperate the coin and tdkitty in to seperate vecs from the output_data .
+    for utxo in output_data {
+        if let Ok(coin) = utxo.extract::<Coin<ID>>() {
+            let utxo_value = coin.0;
+            ensure!(
+                utxo_value > 0,
+                TradeableKittyError::MoneyError(MoneyError::ZeroValueCoin)
+            );
+            output_coin_data.push(utxo.clone());
+            // Process Coin
+        } else if let Ok(td_output_kitty) = utxo.extract::<TradableKittyData>() {
+            // Checking if more than 1 kitty in output is sent for trading.
+            match output_ktty_to_be_traded {
+                None => {
+                    // 1st tradable kitty is received in the output.
+                    output_ktty_to_be_traded = Some(td_output_kitty.clone());
+                    ensure!(
+                        input_ktty_to_be_traded.clone().unwrap().kitty_basic_data
+                            == td_output_kitty.kitty_basic_data, // basic kitty data is unaltered
+                        TradeableKittyError::KittyBasicPropertiesAltered // this need to be chan
+                    );
+                }
+                Some(_) => {
+                    // More than 1 tradable kitty are sent in output for trading.
+                    return Err(TradeableKittyError::CannotBuyMoreThanOneKittyAtTime);
+                }
+            };
+        } else {
+            return Err(TradeableKittyError::BadlyTyped);
+        }
+    }
+
+    // Ensuring we found kitty in the output
+    ensure!(output_ktty_to_be_traded != None, {
+        TradeableKittyError::InputMissingError
+    });
+
+    // Ensuring total money sent is enough to buy the kitty
+    ensure!(
+        total_price_of_kitty <= total_input_amount,
+        TradeableKittyError::InsufficientCollateralToBuyKitty
+    );
+
+    // Filterd coins sent to MoneyConstraintChecker for money validation.
+    MoneyConstraintChecker::<0>::Spend.check(&input_coin_data, &[], &output_coin_data)?;
+    Ok(())
```

Comment:
You might consider returning the difference as a priority. For example if I buy your kitty for 10 coins, but I send a total of 12 coins, the remaining 2 can be a tip to the block author and that should be reflected in the priority. See the money piece to see an example of what I mean.

Reply:
yes now returning the priority as tips .